sycl/include/sycl/ext/oneapi/experimental/sycl_complex.hpp

```diff
@@ -983,6 +983,480 @@ tan(const complex<_Tp> &__x) {
 } // namespace oneapi
 } // namespace ext
 
+template <typename T> using marray_data = sycl::detail::vec_helper<T>;
+
+template <typename T>
+using marray_data_t = typename detail::vec_helper<T>::RetType;
+
+template <typename T, std::size_t NumElements>
+class marray<sycl::ext::oneapi::experimental::complex<T>, NumElements> {
+private:
+  using ComplexDataT = sycl::ext::oneapi::experimental::complex<T>;
+
+public:
+  using value_type = ComplexDataT;
+  using reference = ComplexDataT &;
+  using const_reference = const ComplexDataT &;
+  using iterator = ComplexDataT *;
+  using const_iterator = const ComplexDataT *;
+
+private:
+  value_type MData[NumElements];
+
+  // Trait for checking if an argument type is either convertible to the data
+  // type or an array of types convertible to the data type.
+  template <typename Type>
+  struct IsSuitableArgType : std::is_convertible<Type, ComplexDataT> {};
+  template <typename Type, size_t N>
+  struct IsSuitableArgType<marray<Type, N>>
+      : std::is_convertible<Type, ComplexDataT> {};
+
+  // Trait for computing the conjunction of of IsSuitableArgType. The empty type
+  // list will trivially evaluate to true.
+  template <typename... ArgTN>
+  struct AllSuitableArgTypes : std::conjunction<IsSuitableArgType<ArgTN>...> {};
+
+  // Utility trait for creating an std::array from an marray argument.
+  template <typename ComplexDataT, typename Type, std::size_t... Is>
+  static constexpr std::array<ComplexDataT, sizeof...(Is)>
+  MArrayToArray(const marray<Type, sizeof...(Is)> &A,
+                std::index_sequence<Is...>) {
+    return {static_cast<ComplexDataT>(A.MData[Is])...};
+  }
+  template <typename ComplexDataT, typename Type, std::size_t N>
+  static constexpr std::array<ComplexDataT, N>
+  FlattenMArrayArgHelper(const marray<Type, N> &A) {
+    return MArrayToArray<ComplexDataT>(A, std::make_index_sequence<N>());
+  }
+  template <typename ComplexDataT, typename Type>
+  static constexpr auto FlattenMArrayArgHelper(const Type &A) {
+    return std::array<ComplexDataT, 1>{static_cast<ComplexDataT>(A)};
+  }
+  template <typename ComplexDataT, typename Type> struct FlattenMArrayArg {
+    constexpr auto operator()(const Type &A) const {
+      return FlattenMArrayArgHelper<ComplexDataT>(A);
+    }
+  };
+
+  // Alias for shortening the marray arguments to array converter.
+  template <typename ComplexDataT, typename... ArgTN>
+  using MArrayArgArrayCreator =
+      detail::ArrayCreator<ComplexDataT, FlattenMArrayArg, ArgTN...>;
```

Comment:
Absolutely. Would bringing these helpers outside of the `marray` class and inside the `detail` namespace of the same `sycl/include/sycl/marray.hpp` file, would work for you?

Reply:
Sure!

Reply:
What's the reason for this FIXME you added to your latest commit?

```
// FIXME: Other marray specializations needs to be a friend to access MData.
// If the subscript operator is made constexpr this can be removed.
template <typename Type_, std::size_t NumElements_> friend class marray;
```

I'm asking because, to move the helper traits outside of the class `marray`, and inside the namespace ```sycl::detail``` (since one of the traits is accessing directly MData), I had to use the subscript operator on the `marray` and make it a `constexpr`.

Reply:
You are right, making the subscript operator `constexpr` solves the issue, but that is not actually in line with the specification. I opened an issue about this here: https://github.com/KhronosGroup/SYCL-Docs/issues/367.

As a work-around, the friend here lets us skip the subscript operator and get the data directly. Maybe an option is to move the implementation into a helper struct that is in turn a friend of `marray` until this is resolved?

Reply:
That could be a temporary solution, yes.
I'll give some time to your issue to "grow", and if `constexpr` is soon added to the spec, I'll keep the PR as is, otherwise, I'll introduce a FIXME: temporary solution.

Reply:
The issue is from February, so I don't think it's worth holding our breaths for a resolution soon.

Reply:
I hesitated between a friend struct and adding a "temporary" getter.

The problem with temporary getter is that this public getter might be used by and will be "harder" to remove if codebases depend on it.

Encapsulating the helper functions into a helper struct without renaming symbols and other "big" modifications looked the easiest to transit when the subscript operators will be `constexpr`.